src/Framework/FileClassifier.cs

```diff
@@ -0,0 +1,250 @@
+                        count--;
+                    }
+
+                    return parent?.FullName ?? path;
+                }
+            }
+#endif
+        }
+
+        /// <summary>
+        ///     Shared singleton instance.
+        /// </summary>
+        public static FileClassifier Shared => s_sharedInstance.Value;
+
+        /// <summary>
+        ///     Try add paths found into set of known immutable paths.
+        ///     Project files under any of these folders are considered non-modifiable.
+        /// </summary>
+        /// <remarks>
+        ///     This value is used by <see cref="IsNonModifiable" />.
+        ///     Files in the NuGet package cache are not expected to change over time, once they are created.
+        /// </remarks>
+        /// <remarks>
+        ///     Example value: <c>"C:\Users\myusername\.nuget\;D:\LocalNuGetCache\"</c>
+        /// </remarks>
+        public void RegisterImmutableDirectories(string? nuGetPackageFolders)
+        {
+            if (nuGetPackageFolders?.Length > 0)
+            {
+                string[] folders = nuGetPackageFolders.Split(s_semicolonDelimiter, StringSplitOptions.RemoveEmptyEntries);
+                foreach (string folder in folders)
+                {
+                    RegisterImmutableDirectory(folder);
+                }
+            }
+        }
+
+        private void RegisterImmutableDirectory(string? directory)
+        {
+            if (directory?.Length > 0)
+            {
+                string d = EnsureTrailingSlash(directory);
+
+                if (_knownImmutableDirectories.TryAdd(d, d))
+                {
+                    _knownImmutableDirectoriesSnapshot = new List<string>(_knownImmutableDirectories.Values);
+                }
+            }
+        }
+
+        private static string EnsureTrailingSlash(string fileSpec)
+        {
+            if (fileSpec.Length >= 1)
+            {
+                char lastChar = fileSpec[fileSpec.Length - 1];
+                if (lastChar != Path.DirectorySeparatorChar && lastChar != Path.AltDirectorySeparatorChar)
+                {
+                    fileSpec += Path.DirectorySeparatorChar;
+                }
+            }
+
+            return fileSpec;
+        }
+
+        /// <summary>
+        ///     Gets whether a file is expected to not be modified in place on disk once it has been created.
+        /// </summary>
+        /// <param name="filePath">The path to the file to test.</param>
+        /// <returns><see langword="true" /> if the file is non-modifiable, otherwise <see langword="false" />.</returns>
+        public bool IsNonModifiable(string filePath)
+        {
+            // In order to have allocation-less iteration we can not use nor foreach neither linq.Any.
+            // We shall copy reference of _knownImmutableDirectoriesSnapshot into local variable as otherwise
+            // it could be changed during for loop enumeration by other thread.
```

Comment:
The had different meaning in mind:
1) We can't use a foreach loop or linq.Any as its iterator allocates
2) We have to keep _knownImmutableDirectoriesSnapshot  in local variable so the possible changes of _knownImmutableDirectoriesSnapshot would not be visible in during this method invocation

Reply:
Ah, missed that. How about:

Avoid a foreach loop or linq.Any because they allocate.
Copy _knownImmutableDirectoriesSnapshot into a local variable so other threads can't modify it during enumeration.

?

Reply:
Like it. Done